### PR TITLE
feat: encode urls before http calls 

### LINKED ISSUE #752: EPIC - Encode URLs before http calls
### Description
- We should add a validation on UrlFormatter to encode urls if it's not encoded. 
- Encoding type RFC 3986

### Examples
`/user/Name Of User/` should be converted to `/user/Name%20Of%20User`.
`/myService%20` should be converted to `/myService%2520`
`/name+user/`should be converted to `/name%2Buser/`

### Important points

- Currently, we ensure the same behavior across platforms when handling URL by using a shared file named [UrlBuilderTestSpec](https://github.com/ZupIT/beagle/blob/master/common/tests/UrlBuilderTestSpec.json), that describes _inputs_ and _expected outputs_ of this logic. So, to continue ensuring the same behavior, we need to add these new cases there as well.

### Related issues
- [ ] Web:
- [ ] Android: #761 
- [ ] iOS: #763

### Reference: 
https://www.w3schools.com/tags/ref_urlencode.ASP

### PR DESCRIPTION
**Pull request by jenifersantoszup**
## Description

We should add a validation on UrlFormatter to encode urls if it's not encoded. For this, we are using the encoding type RFC 3986

## Related Issues

Resolves #752 

## Tests

The same tests remained and we only  added new url checks.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [DCO].
- [X] All existing and new tests are passing.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/ZupIT/beagle/issues
[DCO]: https://developercertificate.org/

### CODE REVIEW

**Comment by theffc** [comment]
For iOS: our test function is leaving some hole for potential problems, like not being able to decode data set and still passing validation, handling some edge cases inside it. So I would suggest this modifications there:

```swift
func testUrlBuilderCases() throws {
    let jsonData = try jsonFromFile(fileName: "UrlBuilderTestSpec").data(using: .utf8)!
    let dataSet = try JSONDecoder().decode([UrlBuilderTestHelper].self, from: jsonData)
    XCTAssertFalse(dataSet.isEmpty)
    
    dataSet.forEach { data in
        let base = URL(string: data.base ?? "")
        let builder = UrlBuilder(baseUrl: base)
        let resultUrl = builder.build(path: data.path)
        
        XCTAssertEqual(resultUrl?.absoluteString, data.result)
    }
}
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/networking/urlbuilder/UrlBuilderDefault.kt [hunk 1]
```diff
@@ -16,21 +16,43 @@
...
+            .replace("+", "%20")
+            .replace("%2F", "/")
+            .replace("%25", "%")
+            .replace("%2B", "+")
+            .replace("%23", "#")
+            .replace("%3F", "%C3%A9")
+            .replace("*", "%2A")
+            .replace("%7E", "~")
+            .replace("%3A", ":")
+            .replace("%C3%A9%26q%3Dquery", "?&q=query")
```
Comment (tuliopereirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/networking/urlbuilder/UrlBuilderDefault.kt:49]: I think we should handle each character separately, because if we add different case this code will broke.
Example:
`https://www.test.com/?name=test`

**Comment by theffc** [comment]
It seems like we are creating some weird code to handle this test case:

```json
{
    "base": null,
    "path": "/",
    "result": null
}
```

Android uses this verification in `endpointIsNullAndHasAPath` and `endpointIsNullAndHasNotAPath`, and iOS uses a weird if to handle this `encodedPath == "/" ? nil : URL(string: encodedPath)`.

Since this case is similar to all other cases of a _relative path_ with _no base url_, and with those other cases the expected result is just the actual relative path, I think it would be better to just change the expected result of this case to be `"/"` and remove these weird codes.

Obs: I'm not sure how web is handling this, but maybe there they have a weird if as well.
